src/language-js/printer-estree.js

```diff
@@ -3094,12 +3094,29 @@ function printPathNoParens(path, options, print, args) {
       return "unknown";
     case "TSVoidKeyword":
       return "void";
-    case "TSAsExpression":
-      return concat([
-        path.call(print, "expression"),
-        " as ",
-        path.call(print, "typeAnnotation")
+    case "TSAsExpression": {
+      const typeText = options.originalText.slice(
+        options.locStart(n.typeAnnotation),
+        options.locEnd(n.typeAnnotation)
+      );
+      // Defer to expansion logic for objects, arrays, and annotations wrapped in parentheses.
+      const mightBeMultiline = /^[({[]/.test(typeText.trim());
```

Comment:
No need regexp here, you need check ast nodes inside

Reply:
Is there an easier way to access to the raw string value of the node to check if it’s wrapped in parentheses?

Reply:
You don't need raw, you need look in ast for nodes

Reply:
I don't think the parentheses aren't attached to nodes, though, so we'd have to have access to either the raw text or to iterate over a token list. For the objects and arrays, we definitely can check the AST nodes. Alternatively, we could change the behavior as suggested [above](https://github.com/prettier/prettier/pull/6450#discussion_r320703051) and always add the newline after the `as` operator, and not check this at all.

Reply:
> always add the newline after the as operator

yes, please

Reply:
The downside of always adding the newline after the `as` operator is that instead of

```ts
const value3 = thisIsAReallyLongIdentifier as {
  prop1: string;
  prop2: number;
  prop3: number;
}[];
```
we would have
```ts
const value3 = thisIsAReallyLongIdentifier as 
  {
    prop1: string;
    prop2: number;
    prop3: number;
  }[];
```

Which isn't a style I think is very common.

Reply:
Yep, this doesn't look good.

Reply:
Overall, I expected `as` to behave the same way as other binary operators. But turnes out other operators add even more line breaks:

**Prettier 1.18.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEA3AhgGwFc4BmTAXkxgAsBLNASTQEEAlOA-ATwBloBzBgBMEMWgDNacAE6YA1AB0omTMCUqVAB2kRNARiSY9AGnUbtugEyHLp5eZ2aShkmYC+IYyF1j0yULjSOgDuAAqBCGjIILjYELRCniAARtK4YADWcDAAyprptFD8yDDSxF7UMAC2+ADqdPBo+WBwOZG0YtgdXNFgaFFehWgyMKFp-FW4yOIEw14AVmgAHgBCaZnZObhVcDyFcNOzcAvLOYX8+HAAioQQ8If4cyD50sPS0cm4yVz40EnahRgtQSNGQAA4AAxeCzDWppTTRbRwN7YA5eACOt3gY10URQuDQAFooHA4CJEl5pHBMbQqWNcBMpkgZo9jiBhlVaCUymy0OdLjc7gdmUcvDAvsChKCkLYQKVcLR8OcAMIQKqTaJQaBokCEYYAFS+eJZwzcbiAA)
```jsx
const value3 =
  thisIsAReallyLongIdentifier +
  {
    prop1: 1,
    prop2: 2,
    prop3: 3
  };

```

Reply:
I think it makes sense to treat it the same as other binary operators. I'll explore that and we can see what the team thinks.